### ISSUE TITLE
Proposal: earthlydoc

### ISSUE THREAD

**Issue by vladaionescu** [earthly/earthly#583]
### Proposal

Certain comments with a certain format are interpreted as documentation. These comments can be used to auto-generate help text and documentation.

For example:
```
earth --help +my-target
earth --help github.com/foo/bar+my-target
```

In addition, any target (or [UDC](https://github.com/earthly/earthly/issues/581)) that is imported from a different Earthly project (different Earthfile - either different dir or different repo) **MUST** have an appropriate earthlydoc or otherwise the system returns an error in *strict* mode (`--strict` enabled).

### How it works

* Comments above targets or user-defined commands (UDC) are treated as documentation of that target / UDC. The comment should start with `# <target/udc-name> ….`
* Comments above ARGs are treated as documentation of the respective ARG. It has to start with `# <arg/param-name> …`
* User can run
  ```
  earth --help +target
  ```
  
  or  
  
  ```
  earth --help +COMMAND
  ```
  
  This would print out the earthlydoc of the target or command, together with a table of the ARGs that can be used.

### Complete example

```Dockerfile
# earth builds the main Earthly binary
earth:
    FROM +code
    # GOOS is the target OS
    ARG GOOS=linux
    # GOARCH is the target architecture
    ARG GOARCH=amd64
    RUN go build ...
    SAVE ARTIFACT ./earth
```

```
$ earth --help github.com/earthly/earthly+earth

earth builds the main Earthly binary

Arguments
  GOOS (default=linux) is the target OS
  GOARCH (default=amd64) is the target architecture

Outputs
  Artifacts
    ./earth
```

### Motivation

* Within the same team, better understanding of how to use the build by people who have not written it
* When publishing common commands to be used within a company, or as open-source libraries, give the users an understanding of how the commands are meant to be used
* A signal that the target or UDC is "exported" and part of a build API of that repository. Maintainers should be careful about maintaining compatibility in case it is being used from another repository that they are not aware of.

### Possible future extensions
* A website similar to godoc can be used to present auto-generated documentation for all the projects using Earthfile on GitHub
* IDE features like hover-on functionality can be implemented based on this (eg hover on a command and see its documentation)

### Credit

* @dchw for the idea.

**Comment by alexcb** [comment]
An additional request, is to have a mode that would list all available targets within an Earthfile (along with a summary of their earthlydoc comment).

**Comment by dchw** [comment]
Callers & Callees would be nice to pull out here so you can understand its position within the build graph.

**Comment by wingyplus** [comment]
Or if we use `doc` sub-command instead?

```
earthly doc +target
```

**Comment by bbugh** [comment]
Love this. I think it would be useful to differentiate the earthlydoc format from just a comment, like how JavaScript, Rust, etc use `/**`, C# and Rust use `///`, Rust uses `//!`, python uses `"""` immediately following the definition, etc. 

```Earthfile
## earth builds the main Earthly binary
# NOTE: this generates blah blah blah
earth:
    FROM +code
    ## GOOS is the target OS
    ARG GOOS=linux
    ## GOARCH is the target architecture
	# WARNING: Looks useless but don't remove it
    ARG GOARCH=amd64
    RUN go build ...
    SAVE ARTIFACT ./earth
```

Imagining this would generate the same thing as your example output:

```
$ earth --help github.com/earthly/earthly+earth

earth builds the main Earthly binary

Arguments
  GOOS (default=linux) is the target OS
  GOARCH (default=amd64) is the target architecture

Outputs
  Artifacts
    ./earth
```

**Comment by mwopfner** [comment]
I think using `##` for doc comments is a great idea and aligns with other programming languages. It would also allow to have a different color/highlight in an IDE for such comments.

Maybe we could also integrate #2102 so that it is possible to indicate which targets are for private/public use. The `--strict` than can also check if all `public` targets of the current Earthfile (not starting with `_`) have an appropriate earthlydoc, not just the ones we have imported.

**Comment by nelsam** [comment]
Hey, all, I just wanted to jot down some quick updates about how this will behave when it's merged:

First: the initial PR only handles doc comments on targets. It required some notable parser/lexer changes, so we just wanted to get the base case working first. There will be follow-ups to get it working on `ARG`, `SAVE ARTIFACT`, and `SAVE IMAGE` lines too.

Second, In order to be considered documentation:
1. Comments must be immediately above the element being documented. Empty lines between the comments and the element will indicate that they are not intended as documentation.
2. Comments must start with the name of the element being documented. Comments that begin with any other word (e.g. `# This is...`) will not be considered documentation.
    - This stems from `godoc`, and there's a reasonably good reason behind it: tools like `grep` are used commonly at the CLI to look for logic that does what a developer needs. If the documentation never mentions the _name_ of the thing it's documenting, then those tools become less useful.

Running `earthly doc` will output all documented targets and their documentation.

Running `earthly doc {targetRef}` for a target that has documentation will output its documentation. In follow-up PRs, this will include extra detail about any `ARG`, `SAVE ARTIFACT`, and `SAVE IMAGE` commands that are part of the target, as well as their documentation.

**Comment by mwopfner** [comment]
Hi, `earthly doc` works great in 0.7 but I stumbled about an edge case which could may be improved.
The following Earthfile
```
VERSION 0.7

# all Builds everything
# The output json file looks like:
# {
#     "foo": "bar"
#     "obj": {
#         "ci": "earthly"
#     }
# }
all:
  FROM alpine
  RUN echo "Test"
```
produces:
```
TARGETS:
  +all
    all Builds everything
    The output json file looks like:
    {
    "foo": "bar"
    "obj": {
    "ci": "earthly"
    }
    }
```
where all indentation is removed. (I guess each comment line is striped of all leading white spaces)

**Comment by nelsam** [comment]
Yep, good catch! we should probably read the leading whitespace of the first comment in the block and only trim that amount from following comments.

**Comment by rrjjvv** [comment]
> * This stems from `godoc`, and there's a reasonably good reason behind it: tools like `grep` are used commonly at the CLI to look for logic that does what a developer needs. If the documentation never mentions the _name_ of the thing it's documenting, then those tools become less useful.

That's the first time I had seen a reasonable explanation for that requirement.  Do you feel that applies to the output as well?  I personally find it unhelpful and distracting to have to see it output a second time.  With some (esp. longer) targets, I find it not just unhelpful, but mildly harmful.

So, for explicit example, instead of
```
  +lint-newline-ending
      lint-newline-ending checks that line endings are unix style and that files end
      with a single newline.
```
would you be open to
```
  +lint-newline-ending
      checks that line endings are unix style and that files end
      with a single newline.
```
instead?  I'm "fine" with the requirement that it be duplicated in the source... but hoping you might consider stripping it on output.

(It did occur to me that go programmers are so used to it, they might experience the exact opposite usability issue with my proposal... they're so accustomed to seeing the duplication in source, that it's disorienting _not_ to see it on output as well.  So I may be in the minority, but can't hurt to ask!)

**Comment by nelsam** [comment]
I've been primarily a go programmer since 2013, and I can tell you it's pretty disorienting to not see the name in the doc string 😅 I semi-regularly find it useful for stuff like `earthly doc | grep -i style`, e.g. when looking for a target that might run a formatter but I don't know the exact project-specific term to look for. If the target name isn't part of the doc string, I have a harder time finding it. 

Rough idea: what if we scrap the name if it's on its own line? e.g.

```
# lint-newline-ending:
#
# checks that line endings are unix style and that files end
# with a single newline.
```

That could result in the output you suggested.

**Comment by rrjjvv** [comment]
I'd be perfectly fine having to adopt a special convention, hidden flag, etc.  By the same token, I don't want my desire to get in the way of the system's evolution as a whole (worrying about compatibility, stopping a future feature from being added because this new convention introduces an inconsistency or ambiguity, etc.).  And I wouldn't mind waiting until the design/syntax gets finalized.

Or maybe just to see if my comment gets any 👍🏻 .  If I'm the only one that 'cares', it's not worth it.

**Comment by mwopfner** [comment]
@rrjjvv I agree with you. Having a special convention to hide the name would be great. I see two possible ways:

1. An additional flag for `earthly doc` (e.g. `--no-target-name`). Here the end user decides how to display the documentation.
2. A special syntax in the Earthfile (as @nelsam) proposed. Here the developer who writes the target decides how it should be displayed.

I would go with version 1 because here the end user/developer can decide how to display it and all the documentation would be consistent. The worst thing would be that both variants are mixed in one Earthfile and the documentation ins inconsistent.

**Comment by nelsam** [comment]
Just so you're aware, @mwopfner, this means that the name could _only_ be stripped when you are the person running the command. This would be 100% of the time right now, but consider the potential future where e.g. `pkg.earthly.dev` exists to parse and document public `Earthfile` libraries. If and when that exists, you would have no option to view it on `pkg.go.dev` without the name as part of the doc string.

I think I largely agree with you about [1] being nicer, but I want to make sure that the implication is mentioned "out loud" as it were.

**Comment by rrjjvv** [comment]
Another idea... it's probably more work on the front-end, but probably offers the most flexibility and safety (outside of doing nothing).  Many go-based tools I use utilize [golang templates](https://pkg.go.dev/text/template#pkg-overview) to let end-users control what is output (and how).  If you exposed the relevant tokens from the AST, a user could do something like `earthly doc --template "{{.TargetName}} ({{.Args | len}} args) - {{.Body}}"`.  That would probably be paired with a new config setting like `global.doc_template` (to allow a user to set a global default template).

That would leave you to generate your opinionated output without having to make special accommodations (just the one-time cost of accepting a template and using it).  This could be extended to `earthly ls`, as well for the day that we (hopefully!) get structured logging/output.

I'm actually not a big user of go templates, since most of the tools that offer one to specify templates also allow one to get the output as raw JSON, where I would then use `jq` to accomplish the same result.  But that could be an option as well... it'd be a little simpler implementation.  But an `earthly doc -o json` that output something like
```json
{"target": {
  "name": "...",
  "args": [
    {"name": "foo", "description": "bar"}
  ],
  "saves": [],
  "body_raw": "..."
}
```
would allow that same flexibility.

The main downside for you (outside of the initial implementation) is that the tokens (in the template approach) or JSON structure would become contract.  But I would think that this would be additive in nature (adding artifact info, adding image info, adding secret info, etc.) and I can't think of anything obvious where you'd need to remove or change things.

Just another idea.   It's fairly common (esp. in the go world).  And the potential bonus is that they could be things that could be implemented by others ("if you want it bad enough, we'll take a PR"), since it seems like less of a design issue or technical challenge, but more of a strategic question/decision.  I know that's all easy for me to say from where I sit (and without having peeked at the code). 😄

**Comment by mwopfner** [comment]
@nelsam Yes, if the doc is published online it would always have the name in it. I think this is actually a good thing, because I rather want to have everything (even if it is not exactly the way I find best) the same way rather than a mix of different things.

@rrjjvv I totally agree with your idea to use golang templates. This is used in a lot of other go-based tools and is even better than adding one flag for a special case we have in mind now. It would allow others to format the doc as they think is the best way :wink:.
 
If the normal output (as well as the json output) could also be based on the golang template only one interface has to be maintained. But maybe we should create a new Proposal for a general `--template` flag?

**Comment by nelsam** [comment]
For consistency with `go list`, I think I'd prefer `-f`/`--format`. But yeah, that shouldn't be hard to do. We're not using a template right now, but if we were, our current template would be something like:

```
{{ if .SingleTarget }}
{{.Usage}}
    {{.Docs}}
  {{- if .RequiredArgs }}
  REQUIRED ARGS:
    {{- range .RequiredArgs }}
    {{ .Flag }}{{ if .Default }}={{.Default}}
      {{- if .Docs }}
      {{.Docs}}
      {{- end }}
    {{- end }}
  {{- end }}
  {{- /* repeat for optional args, artifacts, and images */ }}
{{- else }}
TARGETS:
  {{- range .Targets }}
  {{.Usage}}
      {{.Docs}}
  {{- /* also check for the --long flag to see if this should include args, artifacts, and images */ }}
  {{- end}}
{{- end }}
```

Allowing some parts of that template to be replaced (e.g. the `{{.Docs}}` part) shouldn't be too hard.
